numba_scipy/special/signatures.py

```diff
@@ -0,0 +1,391 @@
+    'eval_sh_legendre': [(numba.types.float64, numba.types.float64,), (numba_long, numba.types.float64,)],
+    'exp1': [(numba.types.float64,)],
+    'expi': [(numba.types.float64,)],
+    'expit': [(numba.types.float64,), (numba.types.float32,)],
+    'expm1': [(numba.types.float64,)],
+    'expn': [(numba.types.float64, numba.types.float64,), (numba_long, numba.types.float64,)],
+    'gamma': [(numba.types.float64,)],
+    'hyp0f1': [(numba.types.float64, numba.types.float64,)],
+    'hyp1f1': [(numba.types.float64, numba.types.float64, numba.types.float64,)],
+    'hyp2f1': [(numba.types.float64, numba.types.float64, numba.types.float64, numba.types.float64,)],
+    'iv': [(numba.types.float64, numba.types.float64,)],
+    'ive': [(numba.types.float64, numba.types.float64,)],
+    'jv': [(numba.types.float64, numba.types.float64,)],
+    'jve': [(numba.types.float64, numba.types.float64,)],
+    'kn': [(numba.types.float64, numba.types.float64,), (numba_long, numba.types.float64,)],
+    'kv': [(numba.types.float64, numba.types.float64,)],
+    'kve': [(numba.types.float64, numba.types.float64,)],
+    'log1p': [(numba.types.float64,)],
+    'log_ndtr': [(numba.types.float64,)],
+    'loggamma': [(numba.types.float64,)],
+    'logit': [(numba.types.float64,), (numba.types.float32,)],
+    'nbdtr': [(numba.types.float64, numba.types.float64, numba.types.float64,), (numba_long, numba_long, numba.types.float64,)],
+    'nbdtrc': [(numba.types.float64, numba.types.float64, numba.types.float64,), (numba_long, numba_long, numba.types.float64,)],
+    'nbdtri': [(numba.types.float64, numba.types.float64, numba.types.float64,), (numba_long, numba_long, numba.types.float64,)],
+    'ndtr': [(numba.types.float64,)],
+    'pdtr': [(numba.types.float64, numba.types.float64,), (numba_long, numba.types.float64,)],
+    'pdtrc': [(numba.types.float64, numba.types.float64,), (numba_long, numba.types.float64,)],
+    'pdtri': [(numba.types.float64, numba.types.float64,), (numba_long, numba.types.float64,)],
+    'psi': [(numba.types.float64,)],
+    'rgamma': [(numba.types.float64,)],
+    'smirnov': [(numba.types.float64, numba.types.float64,), (numba_long, numba.types.float64,)],
+    'smirnovi': [(numba.types.float64, numba.types.float64,), (numba_long, numba.types.float64,)],
+    'spence': [(numba.types.float64,)],
+    'xlog1py': [(numba.types.float64, numba.types.float64,)],
+    'xlogy': [(numba.types.float64, numba.types.float64,)],
+    'yn': [(numba.types.float64, numba.types.float64,), (numba_long, numba.types.float64,)],
+    'yv': [(numba.types.float64, numba.types.float64,)],
+    'yve': [(numba.types.float64, numba.types.float64,)]
+}
+
+name_and_types_to_pointer = {
+    ('agm', numba.types.float64, numba.types.float64): ctypes.CFUNCTYPE(ctypes.c_double, ctypes.c_double, ctypes.c_double)(get_cython_function_address('scipy.special.cython_special', 'agm')),
```

Comment:
I've been testing it with this script:

```python
import timeit
import ctypes

import scipy.special as sc
import numba
import numba_scipy.special

expit_address = numba.extending.get_cython_function_address(
    'scipy.special.cython_special',
    '__pyx_fuse_0expit',
)

expit_pointer = ctypes.CFUNCTYPE(
    ctypes.c_double,
    ctypes.c_double,
    ctypes.c_int,
)(expit_address)

expit_pointer_no_dispatch = ctypes.CFUNCTYPE(
    ctypes.c_double,
    ctypes.c_double,
)(expit_address)


@numba.njit
def expit_dispatch_is_1():
    for i in range(10**4):
        expit_pointer(2.0, 1)


@numba.njit
def expit_dispatch_is_0():
    for i in range(10**4):
        expit_pointer(2.0, 0)


@numba.njit
def expit_dispatch_not_specified():
    for i in range(10**4):
        expit_pointer_no_dispatch(2.0)


@numba.njit
def expit_numba_scipy():
    for i in range(10**4):
        sc.expit(2.0)


if __name__ == '__main__':
    print(timeit.timeit(expit_dispatch_is_1, number=100))
    print(timeit.timeit(expit_dispatch_is_0, number=100))
    print(timeit.timeit(expit_dispatch_not_specified, number=100))
    print(timeit.timeit(expit_numba_scipy, number=100))
```

which on my machine gives this:

```
0.07619187600000005
0.039562226000000034
0.036896394
0.049485624000000006
```

So it seems like ignoring it is fine? Not sure why the `numba_scipy` version is consistently a little slower though. Maybe immediately after this I should set up some asv benchmarks so that we can try to nail down what's happening more precisely.

Reply:
I was more worried about correctness, but this indeed is a good point. I think the above code is including compilation time in the timings, hence the first function being quite expensive as numba compilation machinery is initialised for that one, and then the rest benefit from it. The rest of the functions will also contain some compilation cost which is likely why their timings are a bit bumpy. This:
```python
import timeit
import ctypes

import scipy.special as sc
import numba
import numba_scipy.special

expit_address = numba.extending.get_cython_function_address(
    'scipy.special.cython_special',
    '__pyx_fuse_0expit',
)

expit_pointer = ctypes.CFUNCTYPE(
    ctypes.c_double,
    ctypes.c_double,
    ctypes.c_int,
)(expit_address)

expit_pointer_no_dispatch = ctypes.CFUNCTYPE(
    ctypes.c_double,
    ctypes.c_double,
)(expit_address)


@numba.njit
def expit_dispatch_is_1():
    for i in range(10**4):
        expit_pointer(2.0, 1)


@numba.njit
def expit_dispatch_is_0():
    for i in range(10**4):
        expit_pointer(2.0, 0)


@numba.njit
def expit_dispatch_not_specified():
    for i in range(10**4):
        expit_pointer_no_dispatch(2.0)


@numba.njit
def expit_numba_scipy():
    for i in range(10**4):
        sc.expit(2.0)


if __name__ == '__main__':
    # warm up
    expit_dispatch_is_1()
    expit_dispatch_is_0()
    expit_dispatch_not_specified()
    expit_numba_scipy()

    print(timeit.timeit(expit_dispatch_is_1, number=100))
    print(timeit.timeit(expit_dispatch_is_0, number=100))
    print(timeit.timeit(expit_dispatch_not_specified, number=100))
    print(timeit.timeit(expit_numba_scipy, number=100))
```
produces near consistent times for me.